draft-barnes-mls-protocol.md

```diff
@@ -1435,44 +1434,219 @@ all arrive at the following state:
+a one-message-per-application-secret limit.
 
-Possible candidates for that are:
+The next generation of Application Secret is computed by deriving an
+Application_Secret_N+1 from Application_Secret_N as follows:
 
- * the key used for the previous message (hash ratcheting)
- * the counter of the previous message (needs to be known to new members of the group)
- * the hash of the previous message (proof that other participants saw the same history)
+~~~~
+application_secret_N+1 =
+  HKDF-Expand-Label(application_secret_N,"mls app upd","",Hash.length)
+~~~~
 
-The requirement for this is that all participants know these values.
-If additional clear-text fields are attached to messages (like the counter), those
-fields MUST be protected by a signed message envelope.
+The Application context provided together with the previous Application secret
+is used to bind the Application messages with the next key and add some freshness.
+
+[[OPEN ISSUE: Context is left empty for now, it might be that using only
+the message counter is enough, this would be more conveniant.
+Hashing all the data is obviously very costly and prevents from encrypt in
+parallel, an other solution could be to add a pseudo-random contribution to
+each message and only hash these. ]]
+
+### Application AEAD Key Calculation
+
+The Application AEAD keying material is generated from the following
+input values:
+
+- The Application Secret value;
+- A purpose value indicating the specific value being generated;
+- The length of the key being generated.
+
+The traffic keying material is generated from an input traffic secret value using:
+
+~~~~
+write_nonce_[sender] =
+  HKDF-Expand-Label(Application_Secret,"mls app nonce", [sender], nonce_length)
+
+write_key_[sender] =
+  HKDF-Expand-Label(Application_Secret,"mls app key", [sender], key_length)
+~~~~
+
+In this figure, [sender] represents the four-byte value representing the
+participant index in the tree used for the group key establishment mechanism.
+Note, that because the identity of the participant using the keys to send data
+is included in the initial Application Secret, all successive updates to the
+Application secret will implicitely inherit this ownership.
+
+All the traffic keying material is recomputed whenever the underlying
+Application Secret changes.
+
+
+## Message Encryption and Decryption
+
+The Group participants should use the AEAD algorithm associated with
+the negotiated MLS ciphersuite to AEAD encrypt and decrypt their
+Application messages and sign them as follows:
+
+~~~~~
+    struct {
+        uint32 length;
+        opaque content[length];
+        opaque signature[signature_length];
+        uint8 zeros[length_of_padding];
+    } MLSPlaintext;
+
+    struct {
+        opaque group<0..2^32-1>;
+        uint32 epoch_secret;
+        uint32 generation;
+        uint32 sender;
+        uint32 length;
+        opaque encrypted_content[MLSCiphertext.length];
```

Comment:
Typically you would do this as opaque encrypted_content<0..2^32-1>;

Reply:
Fixed, Thanks !